### PR TITLE
Add Scrypt PKDF

### PR DESCRIPTION
**Pull request by shamatar**
Fixes #

Checklist:
- [x] Pull request against a `develop` branch.
- [x] Correct file headers (see CONTRIBUTING.md).
- [x] Formatted with [SwiftFormat](https://github.com/nicklockwood/SwiftFormat).
- [x] Tests added.

Changes proposed in this pull request:
- Add Scrypt PKDF
- Internal rounds to Salsa, no public exposure
- Uses some buffer magic and have performance on par with C implementation from libsodium when build in release

### CODE REVIEW

File: Sources/CryptoSwift/Scrypt.swift [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    enum Error: Swift.Error {
+        case nIsTooLarge
+        case rIsTooLarge
+        case nMustBeAPowerOf2GreaterThan1
+        case invalidInput
+    }
+
+    /// Configuration parameters.
+    private let salt: Array<UInt8>
+    private let password: Array<UInt8>
```
Comment (krzyzanowskim) [inline Sources/CryptoSwift/Scrypt.swift]: use `SecureBytes` to store sensitive information.

File: Tests/Tests/ScryptTests.swift [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        let expected: [UInt8] = Array<UInt8>.init(hex: """
+                70 23 bd cb 3a fd 73 48 46 1c 06 cd 81 fd 38 eb
+                fd a8 fb ba 90 4f 8e 3e a9 b5 43 f6 54 5d a1 f2
+                d5 43 29 55 61 3f 0f cf 62 d4 97 05 24 2a 9a f9
+                e6 1e 85 dc 0d 65 1e 40 df cf 01 7b 45 57 58 87
+""".replacingOccurrences(of: " ", with: "").replacingOccurrences(of: "\n", with: "").replacingOccurrences(of: "\t", with: ""))
+        XCTAssertEqual(derived, expected)
+    }
+    
+    //      Takes too long to run in debug mode!
```
Comment (krzyzanowskim) [inline Tests/Tests/ScryptTests.swift]: use #if DEBUG

**Comment by krzyzanowskim** [comment]
Are we going to merge this one? it's almost ready @shamatar

**Comment by tamulis** [comment]
It would be really nice to see this included. 👍

**Comment by CFitzsimons** [comment]
This would be great to see - is this likely to be available soon?

File: Tests/Tests/ScryptTests.swift (hunk 2 above)
Comment (shamatar) [reply Tests/Tests/ScryptTests.swift]: Allowed to run in Release mode

**Comment by shamatar** [comment]
Included comments from @krzyzanowskim. I'm sorry, was swamped by zkSNARKs

**Comment by krzyzanowskim** [comment]
thanks! merged

**Comment by krzyzanowskim** [comment]
hm... looks like `testScrypt_2` is failing on macOS

```
error: -[Tests.Scrypt testScrypt_2] : XCTAssertEqual failed: ("[40, 240, 63, 204, 81, 250, 95, 148, 174, 250, 175, 118, 225, 116, 83, 97, 174, 125, 185, 166, 196, 85, 165, 164, 167, 199, 210, 253, 140, 15, 211, 198, 152, 236, 162, 118, 122, 106, 193, 252, 9, 64, 202, 142, 241, 155, 106, 51, 99, 55, 177, 175, 92, 148, 192, 2, 88, 84, 44, 18, 107, 63, 203, 17]") is not equal to ("[33, 1, 203, 155, 106, 81, 26, 174, 173, 219, 190, 9, 207, 112, 248, 129, 236, 86, 141, 87, 74, 47, 253, 77, 171, 229, 238, 152, 32, 173, 170, 71, 142, 86, 253, 143, 75, 165, 208, 159, 250, 28, 109, 146, 124, 64, 244, 195, 55, 48, 64, 73, 232, 169, 82, 251, 203, 244, 92, 111, 167, 122, 65, 164]") -
Test Case '-[Tests.Scrypt testScrypt_2]' failed (287.129 seconds).
```

**Comment by shamatar** [comment]
Looks like I’ve set p=16 instead of p=1, set me run on a local machine

**Comment by krzyzanowskim** [comment]
@shamatar did you find any solution?

**Comment by shamatar** [comment]
@krzyzanowskim 

Yes, at some point I’ve wrongly copy-pasted a test vector parameters. For a testScrypt_2 the “p” parameter should be 1, while it was 16 I the PR. You can fix this typo by yourself or or revert & use the latest commit from my branch that basically does the same. 

Sincerely, Alex

**Pull request merged**